src/game.cpp

```diff
@@ -363,72 +442,61 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
 
 	// That didn't work, try to find a pointed at node
 
-	v3s16 pos_i = floatToInt(player_position, BS);
-
-	/*infostream<<"pos_i=("<<pos_i.X<<","<<pos_i.Y<<","<<pos_i.Z<<")"
-			<<std::endl;*/
-
-	s16 a = d;
-	s16 ystart = pos_i.Y - (camera_direction.Y < 0 ? a : 1);
-	s16 zstart = pos_i.Z - (camera_direction.Z < 0 ? a : 1);
-	s16 xstart = pos_i.X - (camera_direction.X < 0 ? a : 1);
-	s16 yend = pos_i.Y + 1 + (camera_direction.Y > 0 ? a : 1);
-	s16 zend = pos_i.Z + (camera_direction.Z > 0 ? a : 1);
-	s16 xend = pos_i.X + (camera_direction.X > 0 ? a : 1);
-
-	// Prevent signed number overflow
-	if (yend == 32767)
-		yend = 32766;
-
-	if (zend == 32767)
-		zend = 32766;
+	v3s16 pointed_pos(0, 0, 0);
 
-	if (xend == 32767)
-		xend = 32766;
+	v3f start = camera_position / BS;
+	v3f end = start + (camera_direction * d);
+
+	Interpolator interpolator(start, end);
+
+	std::vector<v3s16> points;
+	std::vector<aabb3f> boxes;
+	while (!interpolator.IsDone())
+	{
+		points.clear();
+		interpolator.Step(points);
+
+		bool hit = false;
+		for (std::vector<v3s16>::iterator iter = points.begin(); iter != points.end(); ++iter)
+		{
+			v3s16 pos = *iter;
+			MapNode n;
+			bool is_valid_position;
+
+			n = map.getNodeNoEx(pos, &is_valid_position);
+			if (!is_valid_position) {
+				continue;
+			}
+			if (!isPointableNode(n, client, liquids_pointable)) {
+				continue;
+			}
 
-	v3s16 pointed_pos(0, 0, 0);
+			boxes.clear();
+			n.getSelectionBoxes(nodedef, &boxes, getNeighbors(pos, nodedef, &map, n));
 
-	for (s16 y = ystart; y <= yend; y++) {
-		for (s16 z = zstart; z <= zend; z++) {
-			for (s16 x = xstart; x <= xend; x++) {
-				MapNode n;
-				bool is_valid_position;
-				v3s16 p(x, y, z);
+			v3s16 np = pos;
+			v3f npf = intToFloat(np, BS);
+			for (std::vector<aabb3f>::const_iterator i = boxes.begin(); i != boxes.end(); ++i) {
+				aabb3f box = *i;
```

Comment:
If you're asking why I pulled it out into a variable, that's what the old code did.